In [2]:
import numpy as np

In [3]:
np.array([1,2,3])

array([1, 2, 3])

In [5]:
arr1 = np.array(range(10))
arr1

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [7]:
arr2 = np.array([0.1,0.2,0.3])
arr2

array([0.1, 0.2, 0.3])

In [8]:
arr3 = np.array(['a', 'b', 'c'])
arr3

array(['a', 'b', 'c'], dtype='<U1')

In [9]:
arr4 = np.array([1,2,'a', 'b', 0.1,0.2])
arr4

array(['1', '2', 'a', 'b', '0.1', '0.2'], dtype='<U11')

In [10]:
arr5 = np.array([1,2,0.3,0.4])
arr5

array([1. , 2. , 0.3, 0.4])

In [12]:
arr6 = np.array(
    [[1,2,3],
    [4,5,6],
    [7,8,9]]
)
arr6

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

## Array 타입

In [ ]:
## 요소 중에 한 요소라도 str이면 모든 요소가 str로 바뀐다
## 요소 중에 한 요소라도 float이면 모든 요소가 float로 바뀐다

In [13]:
arr1.dtype

dtype('int32')

In [14]:
arr2.dtype

dtype('float64')

In [15]:
arr3.dtype

dtype('<U1')

In [18]:
arr3

array(['a', 'b', 'c'], dtype='<U1')

In [16]:
arr4.dtype

dtype('<U11')

In [17]:
arr4

array(['1', '2', 'a', 'b', '0.1', '0.2'], dtype='<U11')

In [19]:
arr5.dtype

dtype('float64')

In [20]:
arr5

array([1. , 2. , 0.3, 0.4])

In [21]:
arr6.dtype

dtype('int32')

## shape

In [22]:
arr5

array([1. , 2. , 0.3, 0.4])

In [23]:
arr5.shape

(4,)

In [24]:
arr6

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [25]:
arr6.shape

(3, 3)

## arange

In [26]:
arr1 = np.arange(10)
arr1

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [27]:
arr2 = np.arange(3,13)
arr2

array([ 3,  4,  5,  6,  7,  8,  9, 10, 11, 12])

In [30]:
arr3 = np.arange(5.0)
arr3

array([0., 1., 2., 3., 4.])

In [32]:
arr4 = np.arange(3,100,3.0)
arr4

array([ 3.,  6.,  9., 12., 15., 18., 21., 24., 27., 30., 33., 36., 39.,
       42., 45., 48., 51., 54., 57., 60., 63., 66., 69., 72., 75., 78.,
       81., 84., 87., 90., 93., 96., 99.])

## linspace

In [36]:
np.linspace(-5,5,10) ## 균등하게 start이상 staop미만

array([-5.        , -3.88888889, -2.77777778, -1.66666667, -0.55555556,
        0.55555556,  1.66666667,  2.77777778,  3.88888889,  5.        ])

## reshape(배열의 형태 변환)

In [38]:
arr = np.arange(36.0)
arr

array([ 0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10., 11., 12.,
       13., 14., 15., 16., 17., 18., 19., 20., 21., 22., 23., 24., 25.,
       26., 27., 28., 29., 30., 31., 32., 33., 34., 35.])

In [39]:
arr.shape

(36,)

In [42]:
arr2 = arr.reshape(12,3)
arr2

array([[ 0.,  1.,  2.],
       [ 3.,  4.,  5.],
       [ 6.,  7.,  8.],
       [ 9., 10., 11.],
       [12., 13., 14.],
       [15., 16., 17.],
       [18., 19., 20.],
       [21., 22., 23.],
       [24., 25., 26.],
       [27., 28., 29.],
       [30., 31., 32.],
       [33., 34., 35.]])

In [43]:
arr2.shape

(12, 3)

In [44]:
## 3차원
arr3 = arr.reshape(3,4,3)
arr3

array([[[ 0.,  1.,  2.],
        [ 3.,  4.,  5.],
        [ 6.,  7.,  8.],
        [ 9., 10., 11.]],

       [[12., 13., 14.],
        [15., 16., 17.],
        [18., 19., 20.],
        [21., 22., 23.]],

       [[24., 25., 26.],
        [27., 28., 29.],
        [30., 31., 32.],
        [33., 34., 35.]]])

In [46]:
arr3.shape

(3, 4, 3)

In [47]:
arr4 = arr.reshape(3,2,2,3)
arr4

array([[[[ 0.,  1.,  2.],
         [ 3.,  4.,  5.]],

        [[ 6.,  7.,  8.],
         [ 9., 10., 11.]]],


       [[[12., 13., 14.],
         [15., 16., 17.]],

        [[18., 19., 20.],
         [21., 22., 23.]]],


       [[[24., 25., 26.],
         [27., 28., 29.]],

        [[30., 31., 32.],
         [33., 34., 35.]]]])

In [48]:
arr1

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [49]:
arr1.shape = (5,2)
arr1

array([[0, 1],
       [2, 3],
       [4, 5],
       [6, 7],
       [8, 9]])

In [50]:
## -1 shape
print(arr3.shape)
arr3

(3, 4, 3)


array([[[ 0.,  1.,  2.],
        [ 3.,  4.,  5.],
        [ 6.,  7.,  8.],
        [ 9., 10., 11.]],

       [[12., 13., 14.],
        [15., 16., 17.],
        [18., 19., 20.],
        [21., 22., 23.]],

       [[24., 25., 26.],
        [27., 28., 29.],
        [30., 31., 32.],
        [33., 34., 35.]]])

In [51]:
arr3.reshape(-1,12) ## -1: 12에 대해 맞춰줌

array([[ 0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10., 11.],
       [12., 13., 14., 15., 16., 17., 18., 19., 20., 21., 22., 23.],
       [24., 25., 26., 27., 28., 29., 30., 31., 32., 33., 34., 35.]])

In [52]:
arr3.reshape(-1)

array([ 0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10., 11., 12.,
       13., 14., 15., 16., 17., 18., 19., 20., 21., 22., 23., 24., 25.,
       26., 27., 28., 29., 30., 31., 32., 33., 34., 35.])

In [53]:
arr3.reshape(4,-1,3)

array([[[ 0.,  1.,  2.],
        [ 3.,  4.,  5.],
        [ 6.,  7.,  8.]],

       [[ 9., 10., 11.],
        [12., 13., 14.],
        [15., 16., 17.]],

       [[18., 19., 20.],
        [21., 22., 23.],
        [24., 25., 26.]],

       [[27., 28., 29.],
        [30., 31., 32.],
        [33., 34., 35.]]])

## astype - 배열의 dtype 변환

In [54]:
arr1 = np.arange(5)
arr2 = np.arange(0.0,0.5,0.1)
arr1, arr2

(array([0, 1, 2, 3, 4]), array([0. , 0.1, 0.2, 0.3, 0.4]))

In [55]:
print(arr1.dtype)
print(arr2.dtype)

int32
float64


In [57]:
print(arr1.astype(float))
print(arr1.astype(np.float64)) ## 보편적인 방법

[0. 1. 2. 3. 4.]
[0. 1. 2. 3. 4.]


In [58]:
print(arr2.astype(str))
print(arr2.astype(np.str))

['0.0' '0.1' '0.2' '0.30000000000000004' '0.4']
['0.0' '0.1' '0.2' '0.30000000000000004' '0.4']


# random

## rand

In [59]:
np.random.rand(5) ## 0~1

array([0.24880826, 0.88861642, 0.40905252, 0.7044698 , 0.75854067])

In [60]:
np.random.rand(3,3)

array([[0.8142059 , 0.54082133, 0.8603955 ],
       [0.68748772, 0.16205848, 0.1551838 ],
       [0.00107536, 0.62670774, 0.86434341]])

In [61]:
np.random.rand(2,2,2)

array([[[0.797197  , 0.5933162 ],
        [0.35864876, 0.18343727]],

       [[0.70654157, 0.348027  ],
        [0.88643959, 0.74851119]]])

## randint

In [62]:
np.random.randint(-5, 3,(3,3)) ## -5부터 3까지 (3,3)shape으로

array([[-5,  1,  0],
       [-4,  1, -1],
       [-1, -3, -5]])

## randn(표준 정규분포)

In [64]:
np.random.randn(2,5)

array([[-0.16481229,  1.47630193, -2.41230357, -1.98648612,  1.00272214],
       [-0.64587737, -1.70536903,  0.63828832,  1.37735253, -0.85889498]])

## array의 indexing, slicing

In [65]:
arr1 = np.arange(0,100,10.0)
arr1

array([ 0., 10., 20., 30., 40., 50., 60., 70., 80., 90.])

In [68]:
arr1[0], arr1[-1], arr1[5]

(0.0, 90.0, 50.0)

In [70]:
arr1[:], arr1[4:], arr1[1:8]

(array([ 0., 10., 20., 30., 40., 50., 60., 70., 80., 90.]),
 array([40., 50., 60., 70., 80., 90.]),
 array([10., 20., 30., 40., 50., 60., 70.]))

In [71]:
arr1[::2], arr1[::3]

(array([ 0., 20., 40., 60., 80.]), array([ 0., 30., 60., 90.]))

## n-Dim Array indexing, slicing

In [72]:
## 2차원 배열 : 0~23까지 차례로 (8,3)의 배열을 만드세요. dtype = float 
arr2 = np.arange(24.0).reshape(8,3)
arr2

array([[ 0.,  1.,  2.],
       [ 3.,  4.,  5.],
       [ 6.,  7.,  8.],
       [ 9., 10., 11.],
       [12., 13., 14.],
       [15., 16., 17.],
       [18., 19., 20.],
       [21., 22., 23.]])

In [74]:
arr2[0], arr2[4]

(array([0., 1., 2.]), array([12., 13., 14.]))

In [80]:
print(arr2[3][1]) ## 성능이 조금 느리다 , 연산을 두번해서
print(arr2[3, 1]) ## 보편적인 방법

10.0
10.0


In [83]:
print(arr2[3,1:])
print(arr2[3:, 1:])
print(arr2[::2])

[10. 11.]
[[10. 11.]
 [13. 14.]
 [16. 17.]
 [19. 20.]
 [22. 23.]]
[[ 0.  1.  2.]
 [ 6.  7.  8.]
 [12. 13. 14.]
 [18. 19. 20.]]


In [86]:
arr2

array([[ 0.,  1.,  2.],
       [ 3.,  4.,  5.],
       [ 6.,  7.,  8.],
       [ 9., 10., 11.],
       [12., 13., 14.],
       [15., 16., 17.],
       [18., 19., 20.],
       [21., 22., 23.]])

In [92]:
print(arr2[[0,5,3]])
print()
print(arr2[[0,5,3],[1]])

[[ 0.  1.  2.]
 [15. 16. 17.]
 [ 9. 10. 11.]]

[ 1. 16. 10.]


In [94]:
arr2[arr2>10]

array([11., 12., 13., 14., 15., 16., 17., 18., 19., 20., 21., 22., 23.])

In [96]:
arr2[arr2%2==1].reshape(3,4)

array([[ 1.,  3.,  5.,  7.],
       [ 9., 11., 13., 15.],
       [17., 19., 21., 23.]])

## numpy의 상수

In [97]:
arr = np.array([1,2,3,np.nan,np.inf,-np.inf])
print("dtype:", arr.dtype)
arr

dtype: float64


array([  1.,   2.,   3.,  nan,  inf, -inf])

## arr의 연산

In [99]:
arr1 = np.arange(0,10,1)
arr2 = np.arange(1,0,-0.1)
arr1, arr2

(array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]),
 array([1. , 0.9, 0.8, 0.7, 0.6, 0.5, 0.4, 0.3, 0.2, 0.1]))

In [100]:
arr1

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [103]:
## array 와 scalar 연산
arr1 + 1,arr1 + [1,1,1,1,1,1,1,1,1,1] ## 같은 결과

(array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10]),
 array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10]))

In [110]:
print(arr1/2)
print(arr1%2)
print(arr1-2)
print(arr1*2)
print(arr1**2)
print(arr1+2)

[0.  0.5 1.  1.5 2.  2.5 3.  3.5 4.  4.5]
[0 1 0 1 0 1 0 1 0 1]
[-2 -1  0  1  2  3  4  5  6  7]
[ 0  2  4  6  8 10 12 14 16 18]
[ 0  1  4  9 16 25 36 49 64 81]
[ 2  3  4  5  6  7  8  9 10 11]


In [112]:
arr1 + arr2

array([1. , 1.9, 2.8, 3.7, 4.6, 5.5, 6.4, 7.3, 8.2, 9.1])

In [115]:
print(arr1-arr2)
print(arr1*arr2)
print(arr1/arr2)
print(arr1**arr2)
print(arr1%arr2)

[-1.   0.1  1.2  2.3  3.4  4.5  5.6  6.7  7.8  8.9]
[0.  0.9 1.6 2.1 2.4 2.5 2.4 2.1 1.6 0.9]
[ 0.          1.11111111  2.5         4.28571429  6.66666667 10.
 15.         23.33333333 40.         90.        ]
[0.         1.         1.74110113 2.15766928 2.29739671 2.23606798
 2.04767251 1.79278996 1.51571657 1.24573094]
[0.  0.1 0.4 0.2 0.4 0.5 0.4 0.1 0.2 0.1]


In [116]:
arr1 > 5

array([False, False, False, False, False, False,  True,  True,  True,
        True])

In [117]:
arr1 > arr2

array([False,  True,  True,  True,  True,  True,  True,  True,  True,
        True])

## 행렬

In [118]:
## 0행렬
np.zeros(5)

array([0., 0., 0., 0., 0.])

In [119]:
np.zeros((3,3))

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

## 일행렬

In [120]:
np.ones(5)

array([1., 1., 1., 1., 1.])

In [121]:
np.ones((3,3))

array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])

## 단위행렬

In [123]:
np.identity(2)

array([[1., 0.],
       [0., 1.]])

In [124]:
np.identity(5)

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])

In [125]:
np.eye(2)

array([[1., 0.],
       [0., 1.]])

In [126]:
np.eye(5)

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])

## 통계를 위한 연산

In [128]:
arr1 = np.arange(1,11.0).reshape(5,2)
arr1

array([[ 1.,  2.],
       [ 3.,  4.],
       [ 5.,  6.],
       [ 7.,  8.],
       [ 9., 10.]])

## sum

In [130]:
arr1.sum()

55.0

In [134]:
arr1.sum(axis = 0) ## 행단위

array([25., 30.])

In [135]:
arr1.sum(axis = 1) ## 열단위

array([ 3.,  7., 11., 15., 19.])

## mean

In [131]:
arr1.mean()

5.5

In [136]:
arr1.mean(axis = 0)

array([5., 6.])

In [137]:
arr1.mean(axis = 1)

array([1.5, 3.5, 5.5, 7.5, 9.5])

## std

In [132]:
arr1.std()

2.8722813232690143

In [140]:
arr1.std(axis = 0)

array([2.82842712, 2.82842712])

In [141]:
arr1.std(axis = 1)

array([0.5, 0.5, 0.5, 0.5, 0.5])

## var

In [133]:
arr1.var()

8.25

In [138]:
arr1.var(axis = 0)

array([8., 8.])

In [139]:
arr1.var(axis = 1)

array([0.25, 0.25, 0.25, 0.25, 0.25])

## min

In [151]:
print(arr1.min())
print(arr1.min(axis = 0))
print(arr1.min(axis = 1))

1.0
[1. 2.]
[1. 3. 5. 7. 9.]


## max

In [152]:
print(arr1.max())
print(arr1.max(axis = 0))
print(arr1.max(axis = 1))

10.0
[ 9. 10.]
[ 2.  4.  6.  8. 10.]


## cumsum

In [153]:
arr1

array([[ 1.,  2.],
       [ 3.,  4.],
       [ 5.,  6.],
       [ 7.,  8.],
       [ 9., 10.]])

In [154]:
arr1.cumsum()

array([ 1.,  3.,  6., 10., 15., 21., 28., 36., 45., 55.])

In [156]:
arr1.cumsum(axis = 0)

array([[ 1.,  2.],
       [ 4.,  6.],
       [ 9., 12.],
       [16., 20.],
       [25., 30.]])

In [157]:
arr1.cumsum(axis = 1)

array([[ 1.,  3.],
       [ 3.,  7.],
       [ 5., 11.],
       [ 7., 15.],
       [ 9., 19.]])

## cumprod

In [158]:
arr1.cumprod()

array([1.0000e+00, 2.0000e+00, 6.0000e+00, 2.4000e+01, 1.2000e+02,
       7.2000e+02, 5.0400e+03, 4.0320e+04, 3.6288e+05, 3.6288e+06])

In [159]:
arr1.cumprod(axis = 0)

array([[1.00e+00, 2.00e+00],
       [3.00e+00, 8.00e+00],
       [1.50e+01, 4.80e+01],
       [1.05e+02, 3.84e+02],
       [9.45e+02, 3.84e+03]])

In [160]:
arr1.cumprod(axis = 1)

array([[ 1.,  2.],
       [ 3., 12.],
       [ 5., 30.],
       [ 7., 56.],
       [ 9., 90.]])

## sqrt, log, exp

In [143]:
np.sqrt(arr1.var(axis = 1))

array([0.5, 0.5, 0.5, 0.5, 0.5])

In [144]:
np.log(arr1)

array([[0.        , 0.69314718],
       [1.09861229, 1.38629436],
       [1.60943791, 1.79175947],
       [1.94591015, 2.07944154],
       [2.19722458, 2.30258509]])

In [145]:
np.exp(arr1)

array([[2.71828183e+00, 7.38905610e+00],
       [2.00855369e+01, 5.45981500e+01],
       [1.48413159e+02, 4.03428793e+02],
       [1.09663316e+03, 2.98095799e+03],
       [8.10308393e+03, 2.20264658e+04]])

In [147]:
2.5e+3

2500.0

In [148]:
2.5e-3

0.0025

In [162]:
arr1 = np.arange(1,7).astype(np.float64).reshape(2,3)
arr2 = np.arange(6,0,-1.0).astype(np.float64).reshape(3,2)
arr1, arr2

(array([[1., 2., 3.],
        [4., 5., 6.]]),
 array([[6., 5.],
        [4., 3.],
        [2., 1.]]))

In [163]:
arr1.shape, arr2.shape

((2, 3), (3, 2))

In [164]:
arr1 @ arr2 ## 방법1

array([[20., 14.],
       [56., 41.]])

In [166]:
arr1.dot(arr2) ## 방법2 : 보편적인 방법

array([[20., 14.],
       [56., 41.]])

In [165]:
np.matmul(arr1, arr2) ## 방법3

array([[20., 14.],
       [56., 41.]])

## 전치

In [167]:
arr1

array([[1., 2., 3.],
       [4., 5., 6.]])

In [170]:
arr3 = np.arange(6).reshape(2,3)
arr3

array([[0, 1, 2],
       [3, 4, 5]])

In [171]:
arr1.shape, arr3.shape

((2, 3), (2, 3))

In [173]:
arr3.transpose().shape

(3, 2)

In [174]:
arr3.T ## 일반적인 방법

array([[0, 3],
       [1, 4],
       [2, 5]])

In [175]:
arr1.dot(arr3.T)

array([[ 8., 26.],
       [17., 62.]])

In [178]:
arr1 = np.arange(1,5).reshape(2,2)
arr1

array([[1, 2],
       [3, 4]])

In [181]:
arr2 = np.linalg.inv(arr1)
arr2

array([[-2. ,  1. ],
       [ 1.5, -0.5]])

In [183]:
arr2.dot(arr1)

array([[1.0000000e+00, 4.4408921e-16],
       [0.0000000e+00, 1.0000000e+00]])

## concat

In [184]:
arr1 = np.arange(4).reshape(2,2)
arr2 = np.arange(0,40,10).reshape(2,2)
arr1, arr2

(array([[0, 1],
        [2, 3]]),
 array([[ 0, 10],
        [20, 30]]))

In [188]:
np.concatenate((arr1,arr2), axis = 0)

array([[ 0,  1],
       [ 2,  3],
       [ 0, 10],
       [20, 30]])

In [189]:
np.concatenate((arr1,arr2), axis = 1)

array([[ 0,  1,  0, 10],
       [ 2,  3, 20, 30]])

In [190]:
np.c_[arr1, arr2] ## 고급 라이브러리 np.c_

array([[ 0,  1,  0, 10],
       [ 2,  3, 20, 30]])

## 연습문제

1.10개의 0벡터를 만들고 4번째와 5번째 index의 값을 1로 한 후 (5,2)행렬로 만들어라

In [199]:
arr1 = np.zeros(10)
arr1[4:6] = 1
arr1 = arr1.reshape(5,2)
print(arr1)

[[0. 0.]
 [0. 0.]
 [1. 1.]
 [0. 0.]
 [0. 0.]]


2.10에서 30까지 2씩 증가하는 vector를 만들고 다음 list와 요소별 사칙연산을 하여라

In [217]:
arr2 = np.arange(10,30,2)
a = np.arange(1,2,0.1)

print(arr2+a)
print(arr2-a)
print(arr2*a)
print(arr2/a)

[11.  13.1 15.2 17.3 19.4 21.5 23.6 25.7 27.8 29.9]
[ 9.  10.9 12.8 14.7 16.6 18.5 20.4 22.3 24.2 26.1]
[10.  13.2 16.8 20.8 25.2 30.  35.2 40.8 46.8 53.2]
[10.         10.90909091 11.66666667 12.30769231 12.85714286 13.33333333
 13.75       14.11764706 14.44444444 14.73684211]


3.<1에서 100>까지 10개를 균등하게 분할하여 배열을 만들고, 각각의 합, 평균, 표준편차, 분산, 최소값, 최대값, 누적합, 누적곱을 구하여라

In [269]:

arr3 = np.linspace(1,100,10)
arr3
print(arr3.sum())
print(arr3.mean())
print(arr3.std())
print(arr3.var())
print(arr3.min())
print(arr3.max())
print(arr3.cumsum())
print(arr3.cumprod())

array([  1.,  12.,  23.,  34.,  45.,  56.,  67.,  78.,  89., 100.])

4.0에서 10까지 random한 값을 지닌 3x3 실수 행렬을 만들어라

In [281]:
np.random.seed(0)
arr4 = np.random.rand(3,3)
arr4

array([[0.5488135 , 0.71518937, 0.60276338],
       [0.54488318, 0.4236548 , 0.64589411],
       [0.43758721, 0.891773  , 0.96366276]])

5.0에서 10까지 random한 값을 지닌 3x3 정수 행렬을 만들어라

In [224]:
arr5 = np.random.rand(3,3)
arr5

array([[0.17504126, 0.36202989, 0.39930835],
       [0.7262207 , 0.70237552, 0.4211221 ],
       [0.23203492, 0.41268158, 0.71562218]])

6.5x2 행렬을 만들고 Standardization(정규화 하여라)
standardization : (x-mean(x)/std(x))
mean과 std는 Row단위로 연산을 해야함.(return shape이 (2,))
전체 결과는 5X2 행렬

In [230]:
arr6 = np.arange(0,10).reshape(5,2)
mean6 = arr6.mean(axis = 0)
std6 = arr6.std(axis = 0)
result = (arr6-mean6)/std6
print(result)

[[-1.41421356 -1.41421356]
 [-0.70710678 -0.70710678]
 [ 0.          0.        ]
 [ 0.70710678  0.70710678]
 [ 1.41421356  1.41421356]]


In [282]:
arr1 = np.linspace(-1000,1000,5).reshape(5,1)
arr2 = np.linspace(0,10,5)

arr1 = arr1.reshape(5,-1)
arr2 = arr2.reshape(5,-1)
arr3 = np.concatenate((arr1,arr2), axis = 1)
arr3

array([[-1000. ,     0. ],
       [ -500. ,     2.5],
       [    0. ,     5. ],
       [  500. ,     7.5],
       [ 1000. ,    10. ]])

In [290]:
arr4 = (arr3 -arr3.mean(axis = 0))/arr3.std(axis = 0)
arr4

array([[-1.41421356, -1.41421356],
       [-0.70710678, -0.70710678],
       [ 0.        ,  0.        ],
       [ 0.70710678,  0.70710678],
       [ 1.41421356,  1.41421356]])

In [291]:
arr4.mean(), arr4.std(), arr4.var() ## mean = 0, var = 1 표준화

(0.0, 0.9999999999999999, 0.9999999999999998)

7.5x2 행렬을 만들고 Min-Max 정규화 하여라
(x-mean(x))/(max(x)-min(x))
Row단위로 연산.(return shape)
mean과 std는 Row단위로 연산을 해야함.(return shape이(2,))
전체 결과는 5x2행렬

In [233]:
arr7 = np.arange(0,10).reshape(5,2)
mean7 = arr6.mean(axis = 0)
max7= arr6.max(axis = 0)
min7 = arr6.min(axis = 0)

result = (arr7-mean7)/(max7-min7)
print(result)

[[-0.5  -0.5 ]
 [-0.25 -0.25]
 [ 0.    0.  ]
 [ 0.25  0.25]
 [ 0.5   0.5 ]]


In [292]:
arr3

array([[-1000. ,     0. ],
       [ -500. ,     2.5],
       [    0. ,     5. ],
       [  500. ,     7.5],
       [ 1000. ,    10. ]])

In [295]:
(arr3-arr3.min(axis = 0))/(arr3.max(axis = 0)-arr3.min(axis = 0))

array([[0.  , 0.  ],
       [0.25, 0.25],
       [0.5 , 0.5 ],
       [0.75, 0.75],
       [1.  , 1.  ]])